### PR TITLE
- Use staged container builds to improve caching and reduce size

### PR DESCRIPTION
**Pull request by xkortex**
- Image size reduced from 8 GB to just under 1gb. 
- Switched from Make to Ninja for faster builds that do not hog processor
- Removed unneded dependencies
- Added to .dockerignore
- Readme for docker stuff

- Staged Builds
    - Docker's overlay FS means that `rm`ing files does not reduce size
    - Once build artifacts are build, the build dependencies are no longer needed
    - Both of these can be solved by building in a temporary image and copying
    only the needed libraries in
    - Leverages DESTDIR to generate a directory structure that can be just
    copied onto the `/` of the filesystem
    - Similarly, the data files (like models) can be downloaded ahead of time
    into their own image and copied in. This saves on network IO.
    - Anything in a RUN directive that is non-deterministic (e.g. downloading
    a file from a link, the content of that link changes) does not cause a cache
    miss, so if you need to update something RUN uses, either modify the
    dockerfile or build with `--no-cache` to force a rebuild
- Switch to Ninja
    - cmake can generate many types of build systems
    - Ninja builds faster than GNU Make
    - `make -j` has a tendency to lock up my system when building locally
    - Do not need to tell ninja how many jobs to run
- .dockerignore
    - Paths in .dockerignore are basically invisible to dockerd, so when dockerd
    zips up the build context, all of the cruft can be ignored
    - it is beneficial to docker build speed to add any large, unnecssary files
    and directories to .dockerignore.
    - Just remember they cannot be seen by dockerd

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    mkdir -p dlib/build &&\
+    cd dlib/build &&\
+    cmake -DCMAKE_BUILD_TYPE=Release -G Ninja .. &&\
+    ninja && \
+    ninja install && \
+    DESTDIR=/root/diff ninja install &&\
+    ldconfig
+
+## ==================== Building OpenCV ======================
+ENV OPENCV_VERSION=4.0.1
```
Comment (TadasBaltrusaitis) [inline docker/Dockerfile]: Can use OpenCV 4.1.0, I found it to be slightly faster

**Comment by TadasBaltrusaitis** [comment]
This is awesome! I'm sure it will be helpful to a lot of people. Thanks for writing the instructions as well.

File: docker/Dockerfile (hunk 1 above)
Comment (xkortex) [reply docker/Dockerfile]: Yeah, that builds fine!

**Comment by xkortex** [comment]
I just pushed the update to 4.1. I _think_ that should have been edited in the github gui, but I am still learning my way around the in-browser edits. Either way, should be good for merge.

**Pull request merged**